### PR TITLE
Fourth week

### PR DESCRIPTION
**Pull request by Kondratovich**

### CODE REVIEW

File: WebAutopark.Core/Interfaces/Repositories/IOrderItemRepository.cs [hunk 1]
```diff
@@ -0,0 +1,10 @@
+﻿using System.Collections.Generic;
+using WebAutopark.Core.Entities;
+
+namespace WebAutopark.Core.Interfaces.Repositories
+{
+    public interface IOrderItemRepository : IRepository<OrderItem>
+    {
+        public IEnumerable<OrderItem> GetItemsByOrderId(int orderId);
```
Comment (kuzmiich) [inline WebAutopark.Core/Interfaces/Repositories/IOrderItemRepository.cs:8]:
```suggestion
        IEnumerable<OrderItem> GetItemsByOrderId(int orderId);
```
Redundant

File: WebAutopark.Core/Interfaces/Repositories/IOrderRepository.cs [hunk 2]
```diff
@@ -0,0 +1,9 @@
+﻿using WebAutopark.Core.Entities;
+
+namespace WebAutopark.Core.Interfaces.Repositories
+{
+    public interface IOrderRepository : IRepository<Order>
+    {
+        public int CreateAndReturnId(Order item);
```
Comment (kuzmiich) [inline WebAutopark.Core/Interfaces/Repositories/IOrderRepository.cs:7]:
```suggestion
        int CreateAndReturnId(Order item);
```
Redundant

File: WebAutopark.Core/Interfaces/Repositories/IVehicleRepository.cs [hunk 3]
```diff
@@ -0,0 +1,11 @@
+﻿using System.Collections.Generic;
+using WebAutopark.Core.Entities;
+using WebAutopark.Core.Enums;
+
+namespace WebAutopark.Core.Interfaces.Repositories
+{
+    public interface IVehicleRepository : IRepository<Vehicle>
+    {
+        public IEnumerable<Vehicle> GetAllSortedItems(SortState sortOrder);
```
Comment (kuzmiich) [inline WebAutopark.Core/Interfaces/Repositories/IVehicleRepository.cs:9]:
```suggestion
        IEnumerable<Vehicle> GetAllSortedItems(SortState sortOrder);
```
Redundant

File: WebAutopark.DatabaseAccess/Repositories/OrderItemRepository.cs [hunk 4]
```diff
@@ -2,25 +2,39 @@
...
+    public class OrderItemRepository : BaseRepository, IOrderItemRepository
     {
         private const string QueryCreate = "INSERT INTO OrderItems (OrderId, ComponentId, Quantity) " +
                                               "VALUES (@OrderId, @ComponentId, @Quantity)";
 
         private const string QueryDelete = "DELETE FROM OrderItems WHERE OrderItemId = @id";
 
-        private const string QueryGetById = "SELECT * FROM OrderItems WHERE OrderItemId = @id";
+        private const string QueryGetById = "SELECT oi.*, c.ComponentId as CompId, c.Name " +
+                                              "FROM OrderItems oi " +
```
Comment (kuzmiich) [inline WebAutopark.DatabaseAccess/Repositories/OrderItemRepository.cs:17]:
```suggestion
                                              "FROM OrderItems AS oi " +
```

File: WebAutopark.DatabaseAccess/Repositories/OrderItemRepository.cs [hunk 5]
```diff
@@ -2,25 +2,39 @@
...
+                                              "INNER JOIN Components c " +
```
Comment (kuzmiich) [inline WebAutopark.DatabaseAccess/Repositories/OrderItemRepository.cs:18]:
```suggestion
                                              "INNER JOIN Components AS c " +
```

File: WebAutopark.DatabaseAccess/Repositories/OrderItemRepository.cs [hunk 6]
```diff
@@ -2,25 +2,39 @@
...
+                                              "ON oi.ComponentId = c.ComponentId " +
+                                              "WHERE OrderItemId = @id";
 
-        private const string QueryGetAll = "SELECT * FROM OrderItems";
+        private const string QueryGetByOrderId = "SELECT oi.*, c.ComponentId as CompId, c.Name " +
+                                                   "FROM OrderItems oi " +
```
Comment (kuzmiich) [inline WebAutopark.DatabaseAccess/Repositories/OrderItemRepository.cs:23]:
```suggestion
                                                   "FROM OrderItems AS oi " +
```

File: WebAutopark.DatabaseAccess/Repositories/OrderItemRepository.cs [hunk 7]
```diff
@@ -2,25 +2,39 @@
...
+                                                   "INNER JOIN Components c " +
```
Comment (kuzmiich) [inline WebAutopark.DatabaseAccess/Repositories/OrderItemRepository.cs:24]:
```suggestion
                                                   "INNER JOIN Components  AS c " +
```

File: WebAutopark.DatabaseAccess/Repositories/OrderItemRepository.cs [hunk 8]
```diff
@@ -2,25 +2,39 @@
...
+                                                   "ON oi.ComponentId = c.ComponentId " +
+                                                   "WHERE OrderId = @orderId";
+
+        private const string QueryGetAll = "SELECT oi.*, c.ComponentId as CompId, c.Name " +
+                                             "FROM OrderItems oi " +
+                                             "INNER JOIN Components c " +
```
Comment (kuzmiich) [inline WebAutopark.DatabaseAccess/Repositories/OrderItemRepository.cs:30]:
```suggestion
        private const string QueryGetAll = "SELECT oi.*, c.ComponentId as CompId, c.Name " +
                                             "FROM OrderItems AS oi " +
                                             "INNER JOIN Components AS c " +
```

File: WebAutopark.DatabaseAccess/Repositories/OrderItemRepository.cs [hunk 9]
```diff
@@ -2,25 +2,39 @@
...
+                                             "ON oi.ComponentId = c.ComponentId";
```
Comment (kuzmiich) [inline WebAutopark.DatabaseAccess/Repositories/OrderItemRepository.cs:31]:
```suggestion
                                             "ON oi.ComponentId = CompId";
```

File: WebAutopark.DatabaseAccess/Repositories/OrderRepository.cs [hunk 10]
```diff
@@ -2,32 +2,59 @@
...
+        private const string QueryGetAll = "SELECT o.*, v.VehicleId as VehId, v.Model, v.RegistrationNumber " +
+                                             "FROM Orders o " +
+                                             "INNER JOIN Vehicles v " +
+                                             "ON o.VehicleId = v.VehicleId";
 
-        private const string QueryUpdate = "UPDATE Orders SET VehicleId = @VehicleId WHERE OrdersId = @OrdersId";
+        private const string QueryUpdate = "UPDATE Orders " +
+                                             "SET VehicleId = @VehicleId, " +
+                                             "SET Date = @Date" +
+                                             "WHERE OrdersId = @OrdersId";
```
Comment (kuzmiich) [inline WebAutopark.DatabaseAccess/Repositories/OrderRepository.cs:31]: Check `AS` in queries

File: WebAutopark.DatabaseAccess/Repositories/VehicleRepository.cs [hunk 11]
```diff
@@ -2,19 +2,29 @@
...
+                                              "FROM Vehicles v " +
+                                              "INNER JOIN VehicleTypes vt " +
+                                              "ON v.VehicleTypeId = vt.VehicleTypeId " +
+                                              "WHERE VehicleId = @id";
 
-        private const string QueryGetAll = "SELECT * FROM Vehicles";
+        private const string QueryGetAll = "SELECT v.*, vt.VehicleTypeId as TypeId, vt.TypeName, vt.TaxCoefficient " +
+                                              "FROM Vehicles v " +
+                                              "INNER JOIN VehicleTypes vt " +
+                                              "ON v.VehicleTypeId = vt.VehicleTypeId";
```
Comment (kuzmiich) [inline WebAutopark.DatabaseAccess/Repositories/VehicleRepository.cs:27]: Check `AS` before reduction

File: WebAutopark.DatabaseAccess/Repositories/VehicleRepository.cs [hunk 12]
```diff
@@ -39,6 +50,44 @@ public class VehicleRepository : BaseRepository, IRepository<Vehicle>
...
+        {
+            var queryGetAllSortedItems = sortOrder switch
+            {
+                SortState.ModelDesc => $"{QueryGetAll} ORDER BY Model DESC",
+                SortState.VehicleTypeAsc => $"{QueryGetAll} ORDER BY TypeName ASC",
+                SortState.VehicleTypeDesc => $"{QueryGetAll} ORDER BY TypeName DESC",
+                SortState.MileageAsc => $"{QueryGetAll} ORDER BY Mileage ASC",
+                SortState.MileageDesc => $"{QueryGetAll} ORDER BY Mileage DESC",
+                _ => $"{QueryGetAll} ORDER BY Model ASC"
+            };
```
Comment (kuzmiich) [inline WebAutopark.DatabaseAccess/Repositories/VehicleRepository.cs:63]: Create additional method.

File: WebAutopark.Core/Enums/SortState.cs [hunk 13]
```diff
@@ -0,0 +1,12 @@
+﻿namespace WebAutopark.Core.Enums
+{
+    public enum SortState
```
Comment (kuzmiich) [inline WebAutopark.Core/Enums/SortState.cs:3]: Add IdAsc & IdDesc

File: WebAutopark.DatabaseAccess/Repositories/VehicleRepository.cs [hunk 14]
```diff
@@ -39,6 +50,44 @@ public class VehicleRepository : BaseRepository, IRepository<Vehicle>
...
+        public IEnumerable<Vehicle> GetAllSortedItems(SortState sortOrder)
+        {
+            var queryGetAllSortedItems = sortOrder switch
+            {
+                SortState.ModelDesc => $"{QueryGetAll} ORDER BY Model DESC",
+                SortState.VehicleTypeAsc => $"{QueryGetAll} ORDER BY TypeName ASC",
+                SortState.VehicleTypeDesc => $"{QueryGetAll} ORDER BY TypeName DESC",
+                SortState.MileageAsc => $"{QueryGetAll} ORDER BY Mileage ASC",
+                SortState.MileageDesc => $"{QueryGetAll} ORDER BY Mileage DESC",
+                _ => $"{QueryGetAll} ORDER BY Model ASC"
```
Comment (kuzmiich) [inline WebAutopark.DatabaseAccess/Repositories/VehicleRepository.cs:62]: Sort by Id

File: WebAutopark/Controllers/OrderItemController.cs [hunk 15]
```diff
@@ -0,0 +1,75 @@
...
+            ViewBag.OrderId = orderId;
+
+            return View(orderItemViewModels);
+        }
+
+        [HttpGet]
+        public ActionResult Create(int orderId)
+        {
+            OrderItemViewModel viewModel = new OrderItemViewModel() { OrderId = orderId };
+            ViewBag.Components = GetComponentsForSelect();
```
Comment (kuzmiich) [inline WebAutopark/Controllers/OrderItemController.cs:54]: Set up ViewBag.Component in this method.
Also rename it to SetComponentsForSelect();
I think it`s better.

File: WebAutopark/Controllers/VehicleController.cs [hunk 16]
```diff
@@ -37,9 +39,13 @@ private List<SelectListItem> GetVehicleTypesForSelect()
         }
 
         [HttpGet]
-        public ActionResult Index()
+        public ActionResult Index(SortState sortOrder = SortState.ModelAsc)
```
Comment (kuzmiich) [inline WebAutopark/Controllers/VehicleController.cs:42]: By default use sort by Id ascending

File: WebAutopark/Models/OrderViewModel.cs [hunk 17]
```diff
@@ -1,16 +1,13 @@
...
 {
     public class OrderViewModel
     {
         public int OrderId { get; set; }
-
         public int VehicleId { get; set; }
-
-        [Required]
-        [DataType(DataType.Date)]
+        public Vehicle Vehicle { get; set; }
```
Comment (kuzmiich) [inline WebAutopark/Models/OrderViewModel.cs:10]: Add spaces

File: WebAutoprk.BusinessLogic/Services/Interfaces/IOrderService.cs [hunk 18]
```diff
@@ -0,0 +1,10 @@
+﻿using WebAutopark.BusinessLogic.Dto;
+using WebAutopark.Core.Interfaces;
+
+namespace WebAutopark.BusinessLogic.Services.Interfaces
+{
+    public interface IOrderService : IDataService<OrderDto>
+    {
+        public int CreateAndReturnId(OrderDto item);
```
Comment (kuzmiich) [inline WebAutoprk.BusinessLogic/Services/Interfaces/IOrderService.cs:8]:
```suggestion
        int CreateAndReturnId(OrderDto item);
```
Redundant

File: WebAutoprk.BusinessLogic/Services/OrderItemService.cs [hunk 19]
```diff
@@ -0,0 +1,27 @@
...
+            : base(mapper, orderItemRepository)
+        {
+            _orderItemRepository = orderItemRepository;
+        }
+
+        public IEnumerable<OrderItemDto> GetItemsByOrderId(int orderId)
+        {
+            var entities = _orderItemRepository.GetItemsByOrderId(orderId);
+            var dtoItems = _mapper.Map<IEnumerable<OrderItemDto>>(entities);
+            return dtoItems;
```
Comment (kuzmiich) [inline WebAutoprk.BusinessLogic/Services/OrderItemService.cs:24]: Add spaces

File: WebAutoprk.BusinessLogic/Services/OrderService.cs [hunk 20]
```diff
@@ -0,0 +1,27 @@
...
+            : base(mapper, orderRepository)
+        {
+            _orderRepositry = orderRepository;
+        }
+
+        public int CreateAndReturnId(OrderDto item)
+        {
+            var orderEntity = _mapper.Map<Order>(item);
+            var insertedOrderId = _orderRepositry.CreateAndReturnId(orderEntity);
+            return insertedOrderId;
```
Comment (kuzmiich) [inline WebAutoprk.BusinessLogic/Services/OrderService.cs:24]: Add spaces

File: WebAutoprk.BusinessLogic/Services/VehicleService.cs [hunk 21]
```diff
@@ -1,14 +1,28 @@
...
+            : base(mapper, vehicleRepository)
+        { 
+            _vehicleRepository = vehicleRepository;
+        }
+
+        public IEnumerable<VehicleDto> GetAllSortedItems(SortState sortOrder)
+        {
+            var entities = _vehicleRepository.GetAllSortedItems(sortOrder);
+            var dtoItems = _mapper.Map<IEnumerable<VehicleDto>>(entities);
+            return dtoItems;
```
Comment (kuzmiich) [inline WebAutoprk.BusinessLogic/Services/VehicleService.cs:25]: Add spaces

**Review by kuzmiich — CHANGES_REQUESTED** [review]

File: WebAutoprk.BusinessLogic/Services/VehicleService.cs (hunk 21 above)
Comment (Kondratovich) [reply WebAutoprk.BusinessLogic/Services/VehicleService.cs:25]: fixed

File: WebAutoprk.BusinessLogic/Services/OrderService.cs (hunk 20 above)
Comment (Kondratovich) [reply WebAutoprk.BusinessLogic/Services/OrderService.cs:24]: fixed

File: WebAutoprk.BusinessLogic/Services/Interfaces/IOrderService.cs (hunk 18 above)
Comment (Kondratovich) [reply WebAutoprk.BusinessLogic/Services/Interfaces/IOrderService.cs:8]: fixed

File: WebAutopark/Models/OrderViewModel.cs (hunk 17 above)
Comment (Kondratovich) [reply WebAutopark/Models/OrderViewModel.cs:10]: fixed

File: WebAutopark/Controllers/VehicleController.cs (hunk 16 above)
Comment (Kondratovich) [reply WebAutopark/Controllers/VehicleController.cs:42]: fixed

File: WebAutopark.DatabaseAccess/Repositories/VehicleRepository.cs (hunk 14 above)
Comment (Kondratovich) [reply WebAutopark.DatabaseAccess/Repositories/VehicleRepository.cs:62]: fixed

File: WebAutopark.DatabaseAccess/Repositories/VehicleRepository.cs (hunk 12 above)
Comment (Kondratovich) [reply WebAutopark.DatabaseAccess/Repositories/VehicleRepository.cs:63]: fixed

File: WebAutopark.Core/Interfaces/Repositories/IVehicleRepository.cs (hunk 3 above)
Comment (Kondratovich) [reply WebAutopark.Core/Interfaces/Repositories/IVehicleRepository.cs:9]: fixed

**Review by PushMeNow — APPROVED** [review]

**Pull request merged**